### DISCUSSION TITLE
Preparations for the PyTorch v2.0 upgrade

### DISCUSSION THREAD

**Discussion by xianyuanliu** [pykale/pykale#367] [General] [answered]
Documenting issues that may arise in this upgrade. (Continuous updating)

Upgrading to PyTorch v2.x:
1. When using PyTorch v2.x, the PyTorch-Lightning (PL) setting should be upgraded. Our current [setting](https://github.com/pykale/pykale/blob/main/setup.py#L20) is lower than v1.6.5. This version is not compatible with PyTorch v2.0.0, but PL v2.0.x is.

Upgrading to PL v2.x:
1. [Guide to upgrade to PL v2.0](https://lightning.ai/docs/pytorch/stable/upgrade/migration_guide.html).
2. The new trainer does not support `training_epoch_end` to avoid memory issues. The alternative is the `on_train_epoch_end`. Details can be found in [this PR](https://github.com/Lightning-AI/lightning/pull/16520). This change affects some of our pipeline trainers (at least including domain_adapter).
3. Parameter `pl.Trainer(gpus=args.gpus)` seems abandoned. The alternative is `pl.Trainer(accelerator="gpu", devices=args.gpus)` from [choosing-gpu-devices](https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#choosing-gpu-devices). This change affects most examples.

This is in progress.

**Comment by haipinglu** [comment]
Indeed, also consider upgrading the support of py3.10 and drop py3.7.

**Reply by xianyuanliu** [reply]
Sure, PyTorch v2.0 [supports Python >=3.8, <=3.11](https://pytorch.org/blog/deprecation-cuda-python-support/).

**Comment by xianyuanliu** [comment, marked as answer]
Done with #401
